libr/core/linux_heap_glibc_64.c

```diff
@@ -0,0 +1,815 @@
+			r_agraph_add_node (g, node_title, node_data);
+			if (first_node) first_node = false;
+			//r_agraph_add_edge (g, prev_node, chunk_node);
+			break;	
+		} 
+		next_chunk_ref += size_tmp;
+		prev_chunk_ref = next_chunk_ref;
+		r_core_read_at (core, next_chunk_ref, (ut8 *)cnk, sizeof (RHeapChunk64));
+		node_data = r_str_newf ("size: 0x%x\n fd:%"PFMT64x", bk:%"PFMT64x"\n", prev_c->size, prev_c->fd, prev_c->bk) ;
+		chunk_node = r_agraph_add_node (g, node_title, node_data);
+		(first_node) ? first_node = false : r_agraph_add_edge (g, prev_node, chunk_node);
+		prev_node = chunk_node;
+	}
+	top = r_agraph_add_node (g, top_title, top_data);
+	if (!first_node) {
+		r_agraph_add_edge (g, prev_node, top);
+		free (node_data);
+		free (node_title);
+	}
+	r_agraph_print (g);
+
+	free (cnk);
+	free (prev_c);
+	free (top_data);
+	free (top_title);
+}
+
+void print_current_heap_64(RCore *core, RHeap_MallocState64 *main_arena, ut64 *initial_brk) {
+	RListIter *iter;
+	RDebugMap *map;
+	ut64 brk_start, brk_end, size_tmp;
+	ut64 top_size = UT64_MAX;
+	RHeapChunk64 *cnk = R_NEW0 (RHeapChunk64);
+
+	if (!cnk) {
+		eprintf ("Warning: out of memory\n");
+		return;
+	}
+
+	if (!core || !core->dbg || !core->dbg->maps){
+		return;
+	}
+
+	r_debug_map_sync (core->dbg);
+	r_list_foreach (core->dbg->maps, iter, map) {
+		if (strstr (map->name, "heap")) {
+			*initial_brk = (map->addr >> 12) << 12;
+			brk_start = map->addr;
+			brk_end = map->addr_end;
+			break;
+		}
+	}
+
+	if (*initial_brk == UT64_MAX) {
+		eprintf ("No Heap section\n");
+		return;
+	}
+
+	ut64 next_chunk = *initial_brk, prev_chunk = next_chunk;
+	top_size = main_arena->top - brk_start;
+	while (next_chunk != main_arena->top && next_chunk != brk_end) {
+		r_core_read_at (core, next_chunk, (ut8 *)cnk, sizeof (RHeapChunk64));
+		size_tmp = (cnk->size >> 3) << 3;
+		if (top_size != UT64_MAX && (size_tmp > top_size || next_chunk + size_tmp > main_arena->top)) {
+			PRINT_YA ("\n  Malloc chunk @ ");
+			PRINTF_BA ("0x%"PFMT64x" ", next_chunk);
+			PRINT_RA ("[corrupted]\n");
+			PRINTF_RA ("   size:%"PFMT64x"\n   fd:%"PFMT64x", bk:%"PFMT64x"\n", cnk->size, cnk->fd, cnk->bk);
+			break;
+		}
+		PRINT_YA ("  Malloc chunk @ ");
+		PRINTF_BA ("0x%"PFMT64x" ", prev_chunk);
+		next_chunk += size_tmp;
+		prev_chunk = next_chunk;
+		r_core_read_at (core, next_chunk, (ut8 *)cnk, sizeof (RHeapChunk64));
+		(cnk->size % 2 == 0) ? PRINT_GA ("[free]\n") : (next_chunk == brk_end) ? PRINT_GA ("[free]\n") : PRINT_GA ("[allocated]\n");
+	}
+	PRINT_YA ("\n  Top chunk @ ");
+	PRINTF_BA ("0x%"PFMT64x, main_arena->top);
+	PRINT_GA (" - [brk_start: ");
+	PRINTF_BA ("0x%"PFMT64x, brk_start);
+	PRINT_GA (", brk_end: ");
+	PRINTF_BA ("0x%"PFMT64x, brk_end);
+	PRINT_GA ("]\n");
+	r_cons_printf ("\n");
+	free (cnk);
+}
+
+int cmd_dbg_map_heap_glibc_64(RCore *core, const char *input) {
+	RHeap_MallocState64 *main_arena = R_NEW0 (RHeap_MallocState64);
+
+	if (!main_arena) {
```

Comment:
out of memory errors are problematic because when they happen they usually do in loop and its better to just return with NULL, so the user can still use the shell with what was already allocated instead of having lot of memory error messages all the time